src/Npgsql/MetricsReporter.cs

```diff
@@ -234,9 +231,10 @@ static IEnumerable<Measurement<double>> GetPreparedCommandsRatio()
                     All = Interlocked.Exchange(ref reporter._commandCounters.All, default)
                 };
 
-                measurements[i] = new Measurement<double>(
-                    (double)counters.PreparedCommandsStarted / counters.CommandsStarted * 100,
-                    reporter._poolNameTag);
+                var value = (double)counters.PreparedCommandsStarted / counters.CommandsStarted * 100;
+
+                if (double.IsNormal(value))
```

Comment:
Looking at the double code and talking with @tannergooding, I think this would be better as:

```C#
if (double.IsFinite(value))
```

IsNormal doesn't get inlined, and has a [comment](https://github.com/dotnet/runtime/blob/b2538c4488d016efcb9334f940d10c2ff0acc800/src/libraries/System.Private.CoreLib/src/System/Double.cs#L199-L202) on it that it is "should be rarely called". The extra case it checks for is for "subnormal" doubles, which are less precise but still valid doubles. I don't think we care about filtering out "subnormal" values here.

Reply:
The idea here is to avoid both NaN and Infinity (zero / zero, non-zero / zero), do you think it's better to do those checks instead of calling IsNormal()? Given that this is an observable counter, this sort of perf micro-optimization doesn't seem worth it...

Reply:
> The idea here is to avoid both NaN and Infinity

That's exactly what `double.IsFinite(value)` checks for.

> Given that this is an observable counter, this sort of perf micro-optimization doesn't seem worth it...

It's also about readability. I had never even heard of `double.IsNormal` before this and had to look it up (and then ask Tanner exactly what it meant because the API doc isn't great).

Looking even deeper, I think there is a correctness issue as well.

```C#
var preparedCommands = 0;
var commandsStarted = 10;
Console.WriteLine(double.IsNormal((double)preparedCommands / commandsStarted * 100));
```
prints
```
False
```

Reply:
Ok thanks, I should have done more due diligence on this. Will change this to IsFinite.

Reply:
Right.

`IsNaN(x)` and `IsInfinite(x)` are exactly what you think and are super cheap.

`IsFinite(x)` is basically `!IsNaN(x) && !IsInfinite(x)` and is likewise super cheap.

`IsSubnormal(x)` checks if the value is a special value that is near, but not zero. Some APIs use these values to assist with underflow handling and others treat it as `0`. It is more complex and therefore slightly expensive. Because it is still finite, it is expected to be rare to call and only done in specialized edge case handling.

`IsNormal(x)` then checks what is effectively `IsFinite(x) && !IsSubnormal(x)`. Because it has to do the subnormal check, it is likewise expensive.